### ISSUE TITLE
summary

### ISSUE THREAD

**Issue by OnthewayLsd** [OnthewayLsd/test#3]
DRL在环境奖励稀疏情况下的学习
摘要：
    本文档主要总结了DRL在环境奖励稀疏情况下探索学习问题的相关工作，介绍了工作中所采取的实验对象、对实验对象的特点分析、现有的对该问题的研究现状、对论文或Github项目的复现情况、对新思路的尝试和实验结果分析，最后给出了对这些工作的总结。
 
目    录
1 问题描述	2
1.1 奖励稀疏	2
1.2 自由度大	2
1.3 经验的可重用性	3
1.4 状态部分可观	3
1.5 局部任务间依赖关系	3
2 现有研究现状：	3
2.1 DQN网络	3
2.2 Pseudo reward + DQN	3
2.3 Pseudo reward + A3C	4
2.4 Pseudo reward + A3C + OnHighScore	4
3 方法复现	4
3.1 Pixel Joint + A3C	4
3.2 Asynchronous DQN + CTS	4
4 思路尝试	5
4.1 加入遗忘机制	5
4.2 分阶段学习	5
5 总结	7
参考	7

1 问题描述
针对当环境奖励稀疏的情况下，如何使DRL算法进行高效的环境探索和学习。采用的实验平台是OpenAI Gym中的游戏MontezumaRevenge-v0。
 
该问题是Gym环境中比较典型的奖励稀疏的问题之一，它存在几个特点：奖励稀疏、自由度大、体现范化性不明显、状态部分可观测、任务间存在依赖关系。
1.1 奖励稀疏
Agent在不同房间中探索，而每个房间只有很少的几个奖励（或惩罚），且要经过很多步后才会获得，在没有奖惩信号的时候，agent无法有效学习，导致学习困难。
1.2 自由度大
本游戏可分为局部任务和全局任务，其中局部任务即在各个房间内获取较高的得分，全局任务即以合理的局部任务完成顺序，实现最终通过本关（到达最终目标房间，获取最大得分）。有些局部任务间存在前后依赖关系，必须按顺序执行；而有些局部任务可以同时作为后续的可选任务，此时选择不同的局部任务，导致该问题后续发展的自由空间较大，从全局任务完成的角度来看难度和复杂度不同，且可以获得的最终奖励不同。所以，本问题需要Agent在以最优的方式完成各个局部任务的同时，仍然需要探索最优的局部任务完成顺序，最终实现完成全局任务。
1.3 经验的可重用性
大部分情况下所面临的局部任务是不同的，且差距较大，在其他局部任务中学习到的好的行为无法指导当前所面临的新的任务，无法体现策略的范化性，也不容易利用策略的泛化能力进行高效学习。除了个别任务中类似的场景下（如：通过激光、躲避怪兽）可能利用了策略的范化性，其余任务场景下都得重新学习，这也导致了训练时间较长。且最后学出来的策略是一个对各个局部任务过拟合的结果。
1.4 状态部分可观
根据当前显示的图像，只能获取Agent在局部任务中的状态，而无法获取全局任务中的状态，且有些局部任务所显示的图像是一样的，但所处的房间不同（相邻的房间不同，全局状态不同），这些都导致了全局状态的部分客观，无法只根据当前图像学习到有效的全局策略。
1.5 局部任务间依赖关系
    有些局部任务间存在前后依赖关系，如必须完成A，才能去探索并学习任务B。假设现在已经学习到了完成A任务的策略，然后到B任务的环境状态下去探索学习，如果采用同一个策略网络，则在对B的探索学习过程中，策略网络的权值发生变化，导致以前学到的A策略被破坏，当episode结束后，则只能困到A任务再次学习，此时无法对B进行学习。上述过程会往复进行，导致对新任务的学习探索过慢，导致学习效率低。

2 现有研究现状：
（结合论文中方法以及Github上的开源项目, Pseudo reward, density model(pixel density, Context tree switch, pixel cnn, autoencoder(hash))）、
2.1 DQN网络
采用DQN网络自学习玩Atari游戏，在很多游戏中取得了非常好的效果，但是对于奖励稀疏的几个游戏，并没有学习到有效的策略，DQN在MontezumaRevenge中的得分几乎为0。
2.2 Pseudo reward + DQN
针对奖励稀疏情况，人为地加入伪奖励（pseudo reward），用density model判断当前状态被访问的频率，访问频率低的状态对应的pseudo reward高，以此启发agent去探索新的状态，并在探索的过程中，去发现为数不多的真正的环境奖励，然后根据环境奖励学习到如何以较优的方式完成局部任务。论文[1]中，采用DQN并加入pseudo reward，可以实现对15个房间的探索(50 million training steps)，并能够稳定的完成 个房间的任务，稳定的获取 分，且在加长训练时间情况下（100 million training steps），有时能够达到6000分以上，但是无法稳定的学习到该策略。
2.3 Pseudo reward + A3C
相比于A3C，DQN的数据利用率较低，训练时间较长，因此项目[2]尝试将A3C和pseudo reward结合起来，在采用简单的density model（pixel joint）时，利用A3C算法也可以学习到一些有效的策略，由于采用的density model不同，加入的pseudo reward不同，导致agent采取了不同的探索行为，因此完成的局部任务不同于2.2中论文所示agent，得分小于论文[1]所示方法（不太具有可比性，因为它们所探索的局部任务不同），但是采用了A3C加速了agent训练的效率。
2.4 Pseudo reward + A3C + OnHighScore
为了充分利用获取到稀疏奖励的经历，项目[3]中加入了OnHighScore机制，当获得真实环境奖励的时候，延长当前奖励信号的影响，对此经历中更多的过程数据进行学习，相比于2.3中方法，学习有所提升，但是由于采用的是A3C方法，它加入的OnHighScore机制破坏了A3C的on policy的性质，在理论解释上存在问题。
3 方法复现
(Tensorpack(a3c and pixel density, CTS), on-high-score, asynchronous q-learning(cts))、
3.1 Pixel Joint + A3C
参考Github上项目[2]和[3]，将原始A3C算法和Pixel joint density model结合起来，基本能够复现项目描述中的得分情况，得分可以稳定在500分，此后继续延长训练时间，虽偶尔出现更高得分（1400分），但无法学些到稳定的搞好的策略。

3.2 Asynchronous DQN + CTS
用项目[4]中代码，修改experience存放路径后（原来存储路径导致程序崩溃），然后运行，从运行曲线来看，该方法的学习效率较高，在7million时候可以稳定达到400分，在20million时候，会偶尔达到3500分，随着训练的进行，3500分出现的频率增高，最后能够稳定的学习到3500分的策略。
 

4 思路尝试
4.1 加入遗忘机制
通过density model判断状态被访问的频率，然后根据频率设置pseudo reward的大小，在局部任务下，该机制非常有效，可以启发agent比较高效的实现对环境的探索，收集到环境中真实奖励，学习到对于局部任务的有效策略。但是对于全局任务来说，agent需要在各个房间切换，有些已经探索完成的房间，可能仍然需要再次经过该房间，从而实现对房间另一个出口外状态的探索，但是由于density model的限制，导致agent认为该房间的pseudo reward奖励很小，限制了agent对全局信息的探索。因此，尝试加入对状态的遗忘机制，在density model更新的时候，乘上一个衰减系数，以此表示记忆衰减。但是通过实验，效果不理想，变得比加入该机制前效果差。
原因分析：遗忘对局部任务来说是不利的，加入不当则影响了density model加入的初衷，导致无法完成对局部任务的探索和学习；但是根据上面的分析，遗忘对于全局任务的某些情况下是有用的。如果只是通过上述方式直接加入，没有对全局和局部任务情况加入区分，则影响了局部任务的学习，也就不利于全局任务的学习。
4.2 分阶段学习
针对局部任务间前后依赖关系，导致的学习过程中对新任务学习探索不够，进而导致学习效率低的问题，又创建了一个网络N2。在学习过程中，用N1网络进行环境交互，并用A3C算法更新N1，当学习到一个好的网络时，用N2保存当前的N1网络。然后继续学习，N1的网络会被更改。当一个episode结束，agent的状态被重置，这时候如果用N1网络进行环境交互，则无法到达以前通过好的N1网络到达的新的局部任务状态，导致对新局部任务的探索学习太慢，所以当agent的状态被重置后，首先用存储的N2网络与环境进行交互，而N2是对好的N1网络的副本，可以通过N2网络使Agent到达新的局部任务，继续对新任务进行探索学习，减少了在局部任务间的往复，以此提高学习效率。
关键代码-初始Agent状态设置：
 
关键代码-N1网络保存：
 
学习曲线如下：
 
Smoothing: 0.0
方法对比：
（蓝-original_a3c_T5, 紫-original_a3c_T20, 橙-state_reset_a3c_T5:）
 
smoothing: 0.6
蓝色曲线对应的是原始A3C + pixel joint(Tmax = 5)的情况，紫色曲线对应的原始A3C + pixel joint(Tmax = 20)，橙色曲线对应上述提到的分阶段学习方法(Tmax = 5)，从曲线中可以看出，在学习最开始的局部任务时，蓝线和橙线学习效率差不多，但是在后续学习新的任务时，橙线的得分要高于蓝线，说明该方法比原始方法的学习效率提高了，且在60million左右，可以比较稳定的达到2500分，但是当达到2500分后，得分不在提高。
    原因分析：可能是4.1中描述的对全局任务的探索受限，由于density model的影响，限制了agent进一步去经过熟悉的区域，到达新的任务状态。
5 总结
通过对论文的复现以及对自己思路的实验，加入pseudo reward对于局部任务的探索学习非常有效，采用的density model可以是Pixel Joint, Context Tree Switch, Pixel CNN，但是采用不同的density model，只会对局部任务的学习效率带来提升，这对于全局任务的策略学习并不是最关键的一个环节，反而在某种程度上限制了agent对全局任务的探索。加入遗忘机制感觉会减弱density model的限制，但是加入不当会导致局部任务无法有效完成。采用分阶段的学习，可以提高学习效率，但也只是对局部任务的学习，仍然无法从全局上去学习到有效的策略。
MontezumaRevenge的环境探索分两部分：局部任务（房间）的探索、对局部任务完成顺序的探索（全局探索）。其中局部探索现有方法基本上可以实现了，但是对于全局探索，现有方法还做的不够好，它们不能有效协调局部探索和全局探索间的关系，导致无法完成全局任务。

参考
[1] https://arxiv.org/abs/1606.01868 (Unifying Count-Based Exploration and Intrinsic Motivation, DeepMind)
[2] https://github.com/pkumusic/E-DRL
[3] https://github.com/pkumusic/E-DRL/tree/master/C-A3C
[4] https://github.com/steveKapturowski/tensorflow-rl
